File: educational-core/src/com/jetbrains/edu/learning/checker/CheckResult.java
package com.jetbrains.edu.learning.checker;

import com.jetbrains.edu.learning.courseFormat.CheckStatus;
import org.jetbrains.annotations.Nullable;

public class CheckResult {
  public static final CheckResult USE_LOCAL_CHECK = new CheckResult(CheckStatus.Unchecked, "Always use local check");
  public static final CheckResult FAILED_TO_CHECK = new CheckResult(CheckStatus.Unchecked, CheckUtils.FAILED_TO_CHECK_MESSAGE);
  public static final CheckResult LOGIN_NEEDED = new CheckResult(CheckStatus.Unchecked, CheckUtils.LOGIN_NEEDED_MESSAGE);
  public static final CheckResult CONNECTION_FAILED = new CheckResult(CheckStatus.Unchecked, "Connection failed");

  private CheckStatus myStatus;
  private String myMessage;
  @Nullable private String myDetails;

  public CheckResult(CheckStatus status, String message) {
    this(status, message, null);
  }

  public CheckResult(CheckStatus status, String message, @Nullable String details) {
    myStatus = status;
    myMessage = message;
    myDetails = details;
  }

  public CheckStatus getStatus() {
    return myStatus;
  }

  public String getMessage() {
    return myMessage;
  }

  @Nullable
  public String getDetails() {
    return myDetails;
  }
}

File: educational-core/src/com/jetbrains/edu/learning/checker/CheckResult.kt
package com.jetbrains.edu.learning.checker

import com.jetbrains.edu.learning.EduNames.FAILED_TO_CHECK_URL
import com.jetbrains.edu.learning.EduNames.NO_TESTS_URL
import com.jetbrains.edu.learning.courseFormat.CheckStatus
import com.jetbrains.edu.learning.messages.EduCoreBundle
import org.jetbrains.annotations.Nls
import javax.swing.event.HyperlinkListener

data class CheckResult @JvmOverloads constructor(
  val status: CheckStatus,
  @Nls val message: String = "",
  val details: String? = null,
  val diff: CheckResultDiff? = null,
  val severity: CheckResultSeverity = CheckResultSeverity.Info,
  val hyperlinkListener: HyperlinkListener? = null
) {

  val fullMessage: String get() = if (details == null) message else "$message\n\n$details"

  val isSolved: Boolean get() = status == CheckStatus.Solved

  companion object {
    @JvmField
    val NO_LOCAL_CHECK = CheckResult(CheckStatus.Unchecked, EduCoreBundle.message("check.result.local.check.unavailable"))
    @JvmField
    val LOGIN_NEEDED = CheckResult(CheckStatus.Unchecked, CheckUtils.LOGIN_NEEDED_MESSAGE)
    @JvmField
    val CONNECTION_FAILED = CheckResult(CheckStatus.Unchecked, EduCoreBundle.message("check.result.connection.failed"))
    @JvmField
    val SOLVED = CheckResult(CheckStatus.Solved)
    @JvmField
    val CANCELED = CheckResult(CheckStatus.Unchecked, EduCoreBundle.message("check.result.canceled"))
    @JvmField
    val UNCHECKED = CheckResult(CheckStatus.Unchecked)

    val noTestsRun: CheckResult
      get() = CheckResult(
        CheckStatus.Unchecked,
        EduCoreBundle.message("gluing.dot", EduCoreBundle.message("check.no.tests"),
                              EduCoreBundle.message("help.use.guide", NO_TESTS_URL))
      )

    @JvmStatic
    val failedToCheck: CheckResult
      get() = CheckResult(
        CheckStatus.Unchecked,
        EduCoreBundle.message("gluing.dot", EduCoreBundle.message("error.failed.to.launch.checking"),
                              EduCoreBundle.message("help.use.guide", FAILED_TO_CHECK_URL))
      )
  }
}

data class CheckResultDiff(val expected: String, val actual: String, val title: String = "")
